core/domain/exp_jobs_one_off.py

```diff
@@ -366,24 +375,33 @@ def map(item):
 
                         new_answer_groups.append(new_answer_group.to_dict())
 
-                    exp_services.update_exploration(
-                        'admin', item.id, [exp_domain.ExplorationChange({
-                            'cmd': 'edit_state_property',
-                            'state_name': state_name,
-                            'property_name': 'widget_id',
-                            'new_value': new_interaction_id
-                        }), exp_domain.ExplorationChange({
-                            'cmd': 'edit_state_property',
-                            'state_name': state_name,
-                            'property_name': 'answer_groups',
-                            'new_value': new_answer_groups
-                        })], 'Upgraded interaction to %s.' % new_interaction_id)
+                    if new_interaction_id != '':
+                        exp_services.update_exploration(
+                            committer_id, item.id, [
+                                exp_domain.ExplorationChange({
+                                    'cmd': 'edit_state_property',
+                                    'state_name': state_name,
+                                    'property_name': 'widget_id',
+                                    'new_value': new_interaction_id
+                                }), exp_domain.ExplorationChange({
+                                    'cmd': 'edit_state_property',
+                                    'state_name': state_name,
+                                    'property_name': 'answer_groups',
+                                    'new_value': new_answer_groups
+                                })], 'Upgraded interaction to %s.' % (
```

Comment:
The thing is, it is not possible to change the type of interaction through the editor UI. The only way to replicate this behavior through the UI is to delete the state and create a new one with a different interaction. I did that and the changes are as expected.
![changelist](https://user-images.githubusercontent.com/35144226/87243890-5249f980-c457-11ea-9b7f-a5cd1e4396d2.gif)

Reply:
What you are doing seems like the right way to achieve 'change interaction' in the frontend. Though, can you maybe log what is present in `UndoRedoService.getCommittableChangeList` or the change list received by the controller  in the backend, when you do 'Delete Interaction -> New Interaction -> Change rules in answer groups'? Since, that is what you are doing, right?

Reply:
+1 -- what @aks681 explains is exactly the evidence I was looking for.

Reply:
@aks681 @seanlip Here are the logs that show the change list when I do the process of creating a new interaction with math expression input -> deleting the state -> creating a new state with Algebraic Expression Input with the same rule input.

![1](https://user-images.githubusercontent.com/35144226/87301799-cfe03900-c52d-11ea-9509-7124feb98377.png)
![2](https://user-images.githubusercontent.com/35144226/87301794-ceaf0c00-c52d-11ea-9f6f-75a5e9ef2a36.png)

Note that the migration job doesn't need to do this (deleting the state and creating a new one), the migration job can directly change the interaction ID which is why the exact changelist in the job is different but logically, they are both the same.

Reply:
You keep talking about "deleting the state". Don't you mean "deleting the interaction"?

Also, please replicate exactly this changelist in your PR. Mimic what would happen in the frontend if you were making the changes manually. If we ever need to do any investigation retroactively in the future, then having everything follow a similar format would be helpful. It would make it easier to assume that the data we have is in a certain format.

Please fix this along the above lines and provide proof that it is done correctly (e.g. for adding the rules etc.). It is important that the backend commit cmd list match exactly whatever would have been done in the frontend had the same commands been performed manually.

Reply:
> You keep talking about "deleting the state". Don't you mean "deleting the interaction"?

No, I do mean "deleting the state". Each interaction has multiple states and each state can have a single interaction with multiple answer groups. So, this job checks for all inputs in all answer groups, and only if all of the inputs are valid and correspond to a single interaction, do we upgrade that state. The point is that states are independent from each other so invalidity of one state should not restrict the upgrade of the other just because they happen to be in the same exploration. So, in a way, I think "deleting the interaction" is the same is "deleting the state" since each state can have exactly one interaction.

> Also, please replicate exactly this changelist in your PR. Mimic what would happen in the frontend if you were making the changes manually. If we ever need to do any investigation retroactively in the future, then having everything follow a similar format would be helpful. It would make it easier to assume that the data we have is in a certain format.

Done. Now, the upgrade job will delete the state and replace it with a copy that has the changes -- just like it would happen in the frontend. I have simply updated the changelist that we update the interaction to match with the one that is generated when the operation is performed from the frontend as shown in the screenshot above.

Reply:
@seanlip, I actually meant "Each _exploration_ (not interaction) has multiple states...". 
And yes, I do mean deleting the interaction and not the state. I was confused between states and interactions.

/cc @aks681